### DISCUSSION TITLE
Default merge strategy (per org, repo or branch)

### DISCUSSION THREAD

**Discussion by Snailedlt** [community/community#20583] [Pull Requests]
# Default Merge Strategy

## Why we need it
In our GitHub organization we have a workflow where the preferred merge strategy is `Squash and merge`. Whenever we merge into master, we usually use this strategy. We would like to be able to set this as the default merge strategy, without removing the options for `Create a merge commit` and `Rebase and merge` for special occasions. Currently the only way to set `Squash and merge` is to disable the other options. This is problematic, since we sometimes need to merge using other merge strategies in special cases, for example if the PR closes multiple issues, and we want one commit per issue it solves. Another thing that makes this hard is that currently GitHub selects the merge strategy based on what you selected ln your last PR. This means that when an unusual option is selected, the user needs to remember to change it back to the "default" so that he/she doesn't accidentally merge with the wrong strategy on their next merge. This is a big source of user errors, and can lead to multiple hours spent trying to fix the (sometimes irreversible) mistakes done by choosing the wrong merge strategy. Instead of having the default be the last selected strategy, it should be possible for repository admins to select which merge strategy is defaulted to.

## Suggested Implementations

1. Add an option in repository settings to choose the default merge strategy.
2. Add an option in branch resteictions, to choose the default merge strategy.
3. Add an option in the GitHub organization settings to choose the default merge strategy for any newly created repositories.

Additionally, if the GitHub UX team finds it useful, it would be nice to have the default merge strategy highlighted in the merge button dropdown. Either with a different background color, or an associated emoji to show if the merge strategy is a default in the org, repo or branch.

### Relevant Links
- #8196
- #13991
- [GitHub Community“Squash and Merge” as default Pull Request merge option](https://github.community/t/squash-and-merge-as-default-pull-request-merge-option/1031)
- [StackOverflow | Is there a way to specify a default option for merging a pull request in GitHub?](https://stackoverflow.com/q/49203040/12206312)

### Implementations by GitHub alternatives

[Bitbucket Cloud](https://bitbucket.org/blog/fast-forward-merges-bitbucket-cloud-default-like) (My favourite)
<img width="826" alt="Screen-Shot-2018-06-06-at-3 53 53-PM" src="https://user-images.githubusercontent.com/43886029/179179198-63e375b8-462b-4a99-ba42-11f8cfe17322.png">

[Bitbucket Server](https://confluence.atlassian.com/bitbucketserver/pull-request-merge-strategies-844499235.html#Pullrequestmergestrategies-Configuremergestrategiesforanindividualrepository)
<img width="698" alt="Screen-Shot-2018-06-06-at-3 57 11-PM" src="https://user-images.githubusercontent.com/43886029/179179244-a48102b9-a6c9-430b-85ef-18a46f758238.png">

[Gitlab](https://gitlab.com/gitlab-org/gitlab/-/issues/17613)
![image](https://user-images.githubusercontent.com/43886029/179180180-ee4869c2-0b8c-4709-beda-b2689047df0a.png)

[Gitea](https://github.com/go-gitea/gitea/issues/17741)
![142775768-97a8368f-531e-4b71-9ef6-39c56147f9f9](https://user-images.githubusercontent.com/43886029/179187816-20a55536-26be-44af-84f2-6d2e51df00b8.png)

## Edit
A partial solution involving the ability to disable certain merge methods has been implemented using repository rules. The solution can be found here: https://github.com/orgs/community/discussions/146284, and is further discussed in this comment: https://github.com/orgs/community/discussions/20583#discussioncomment-11274521

Support for choosing a default merge method per branch still remains to be implemented.

**Comment by tyge68** [comment]
Make totally sense, +1 for such improvement

**Comment by gCardinal** [comment]
It would also be interesting to have the option for users to define a default strategy.

For example, if I work in multiple organisations and Org A defines "Fast Forward" as a default and Org B defines no default, if I set my default to "Squash" merging a PR in Org A would use "Fast Forward" as the strategy whereas merging a PR in Org B would use "Squash".

**Reply by Snailedlt** [reply]
Yes! Definitely would be useful to have default strategies per organization

**Reply by Myera2411** [reply]
> Yes! Definitely would be useful to have default strategies per organization

**Comment by eyadmba** [comment]
A default strategy per branch is a good improvement, but I'd rather completely disable certain strategies with certain branches.

**Reply by Snailedlt** [reply]
That would be a great option too, but it doesn't replace the need for default merge strategies, as in some cases you want to squash merge, and in others you may want to rebase and merge, or merge normally

**Comment by katylava** [comment]
It's pretty annoying to not be able to have different merge strategies per branch. We have repos where we use a strategy of merging PRs into a develop branch, and to release we merge develop into main. We want to enforce squash merge from PRs into develop, but never squash merge from develop to main because then git gets very confused about what has already been merged.

**Reply by katylava** [reply]
Just learned this specific use case is covered in #8196. Be sure to upvote it if this is what you need as well!

**Comment by mskreativ** [comment]
+1

I need to define allowed and default merge method individually for some named branches.

**Comment by sobolk** [comment]
+1

**Comment by rahul1** [comment]
+1 - Very useful for open-source projects with multiple repos

**Comment by ivanwidj** [comment]
+1 - very useful for project with env specific branches

**Comment by ACBatconsumeraffairs** [comment]
+1 This would be very helpful to companies using different merging for lower environments than those used in the upper envs.

**Comment by shaiboujuappcharge** [comment]
+ 1

**Comment by gunnsth** [comment]
+1.  This is absolutely necessary. If I have master, release and develop.  Then I only want to allow squash merge into develop (I usually like 1 commit per change rather than 50 or more, which gets messy imo).  Rest needs to be regular commit.  Otherwise the github compare mechanisms starting thinking there are missing commits and it does not work well.

**Comment by vinod-soni-microsoft** [comment]
+1

**Comment by ConnorCallison** [comment]
+1

**Comment by hq6** [comment]
+1

**Comment by fredericrous** [comment]
other relevant issue that is very similar to this one 👉 https://github.com/orgs/community/discussions/10809

**Reply by Snailedlt** [reply]
Let's link the two together: #10809

**Comment by lyonssp** [comment]
> +1. This is absolutely necessary. If I have master, release and develop. Then I only want to allow squash merge into develop (I usually like 1 commit per change rather than 50 or more, which gets messy imo). Rest needs to be regular commit. Otherwise the github compare mechanisms starting thinking there are missing commits and it does not work well.

I echo this exact value prop

**Comment by SergeyBarskiyTyler** [comment]
+1 👍

**Comment by JoeNunnelley** [comment]
👍

**Comment by Snailedlt** [comment]
It's been almost a year now, and no response.
@github can we bring some attention to this please? 😊

**Comment by guylevy** [comment]
really needed for enterprises! 
+1 👍

**Comment by sudoFerraz** [comment]
+1 👍

**Comment by mamdouni** [comment]
+1

**Comment by ChelseaMD** [comment]
+1

**Comment by vipierozan99** [comment]
+1

**Comment by Measurity** [comment]
> Instead of having the default be the last selected strategy, it should be possible for repository admins to select which merge strategy is defaulted to.

I would also like an option to set a repository as **"always ask when you click the merge button"** since I want to decide on a per PR basis. If a PR has nice commits (separation of concern) I want to merge commits as-is. If it's a mess but the functionality is OK, I want to squash.

**Comment by pgumiela** [comment]
+1

**Comment by johnbaldwin** [comment]
This would be super helpful +1

**Comment by gbartyzel** [comment]
+1

**Comment by sarath-nair-sonarsource** [comment]
+1

**Comment by fruitriin** [comment]
+1

**Comment by rocklan** [comment]
+1

**Comment by chriskessel** [comment]
+1

**Comment by SankarshanDudhate** [comment]
+1. Been waiting for this one since a long time.

**Reply by vintagesucks** [reply]
We all want this feature, but the "+1" messages bother those of us who have subscribed to this thread to catch meaningful updates.

[GitHub Community Code of Conduct](https://docs.github.com/en/site-policy/github-terms/github-community-code-of-conduct#best-practices-for-maintaining-a-strong-community):
> Rather than posting “+1” or “Agreed”, use the upvote button.

**Reply by lobsterkatie** [reply]
Feature request for disallowing (or at least not notifying on) +1 comments:

https://github.com/orgs/community/discussions/53075

**Comment by SankarshanDudhate** [comment]
Also, why is this still in discussion? Can anyone from GitHub team comment on why this is not being implemented?

**Reply by SankarshanDudhate** [reply]
PS: "This is not a priority" or "We're not gonna implement this" are totally acceptable answers here. I only expect some clarity.

**Reply by Snailedlt** [reply]
@github can we get some clarity here?

**Comment by starryeyez024** [comment]
This has serious impacts to a project's commit history, namely inadvertently erasing important information around related issues that were addressed in that batch of code when something is accidentally squashed. This really should be addressed immediately, either by allowing users the ability enable squashing abilities but turn it off as the default, or by [protecting some branches from squashing](https://github.com/orgs/community/discussions/10809).

**Reply by Snailedlt** [reply]
Absolutely!
Being able to set default merge strategy and allowed merge strategies for each branch would solve this issue

**Comment by vintagesucks** [comment]
I just realized that one of the rules I need (only allow "Squash and merge" in a protected branch but allow merge commits generally) is already possible and looks like this for a PR targeting the protected branch:

<img width="419" alt="grafik" src="https://github.com/community/community/assets/13335308/5ddf2610-c63a-4e93-9fdc-7e264e940210">

^ _Not enabled for this branch_

---

Relevant settings:

<details>
  <summary>Repository</summary>
  <img width="822" alt="grafik" src="https://github.com/community/community/assets/13335308/7f828041-7cc7-4292-8040-813ecae9d7bd">
</details>

<details>
  <summary>Protected branch rule</summary>
  <img width="786" alt="grafik" src="https://github.com/community/community/assets/13335308/4ba727b1-9a83-40ea-92c4-b4bfea883cdd">
</details>

The "Require linear history" part was new to me and ensures that only squashing is allowed in my protected branch, although normal merges in the repository are not disabled.

(I enabled "Do not allow bypassing the above settings" here to enforce this for admins [me] as well)

**Reply by katylava** [reply]
Hah, I just saw that yesterday too. Solves half the problem.

**Reply by katylava** [reply]
If only they would add a "Require merge commits" option as well, then I could set that up for the main branch, and "Require linear history" for the develop branch, and the whole problem would be solved.

**Comment by palesnia** [comment]
+1

**Comment by annepruner** [comment]
+1. I just went looking for this because I forget to choose and end up just merging.

**Comment by facundodartayete** [comment]
+1

**Comment by Alucido** [comment]
+1

**Comment by SeanFeldman** [comment]
Lack of support for the per-branch merging strategy is causing unnecessary friction and pain. Please add the feature.

**Comment by Kai-Zhang** [comment]
+1. It's super useful when you manage a repo and have been fixing other's misoperation from time to time. I like the "last operation as default" choice, but sometimes one exception leads to sequential misoperations.

**Comment by darwin808** [comment]
+1 need different strategy per branch

**Comment by SeanFeldman** [comment]
Not a single week where a human error (and the default UI remembering the previous selection) is causing this issue 😞
This feature is needed like yesterday.

**Comment by acman** [comment]
+1 really need it

**Comment by BFS-JWesseler** [comment]
+1 Please implement it.

**Comment by AneeshChuttani** [comment]
+1, this feature will be very helpful.

**Comment by kevin-presalytics** [comment]
+1 Would love this feature!

**Comment by EtiennePASSOT** [comment]
+1

**Comment by TenetMax** [comment]
+1

**Comment by dev-nicolaos** [comment]
Could those who are coming here to comment "+1" or similar just upvote the original post instead? Commenting sends a notification to everyone who is subscribed to this thread (which is a lot of people at this point) and doesn't really add any value to the discussion. Plus if we all upvote that lets people better understand the scale of people wanting this at a glance rather than having to scroll through a long list of repetitive comments.

**Comment by SeanFeldman** [comment]
It looks like people who are "+1-ing" are also upvoting. The count is constantly going up.

Perhaps **someone at GitHub** should seriously re-evaluate the stance on this feature request and do something about it. It's a constant source of friction having to manually remember what merging strategy to apply based on the branch.

**Comment by sunho-cho** [comment]
+1

**Comment by aberenda-optifino** [comment]
+1

**Comment by nader-zouaoui** [comment]
+1

**Comment by abhishekzepto** [comment]
+1

**Comment by nivivive** [comment]
+1

**Comment by cantecim** [comment]
+1, still no answer on this?

**Comment by koenigt** [comment]
+1

**Comment by vintagesucks** [comment]
**Please stop posting "+1" comments.**

I've already mentioned this in [a comment](https://github.com/orgs/community/discussions/20583#discussioncomment-6719713) above, but in the hope that this will reach more people, I'll quote the [GitHub Community Code of Conduct](https://docs.github.com/en/site-policy/github-terms/github-community-code-of-conduct#best-practices-for-maintaining-a-strong-community) again:

> Rather than posting “+1” or “Agreed”, use the upvote button.

**Reply by katylava** [reply]
the only people who are going to see this are the people who have already posted +1 comments. the people who haven't done so yet aren't going to read your message about not doing that... they are going to skip to the bottom to add a comment, especially on a long discussion like this one. so, your comment about "please stop posting +1 comments" is just as pointless as the comments you are complaining about. for the rest of us, it causes the same problem as a +1 comment. 

i debated posting this reply for the same reason, but at least there is a chance the people who get this notification will not post again.

**Comment by SeanFeldman** [comment]
@vintagesucks, you're correct. At the same time, rules are recommendations if not enforced.

**Reply by vintagesucks** [reply]
I know, but at the moment I am more annoyed by daily "+1" comments in several discussions I follow than by missing features. Worth trying, I suppose. 😅

By the way: You can write a direct reply below an answer instead of creating a new one, as you did here.

**Reply by SeanFeldman** [reply]
Thanks, something new for my toolbox 🙂

**Reply by cantecim** [reply]
@vintagesucks you can adjust your subscription settings to this discussion. Everyone might be using both upvoting and +1 too. You can't prevent that by one self explanatory comment. If you are annoyed, solve it by your hands, you'll get no result otherwise

**Reply by vintagesucks** [reply]
@cantecim When you say "adjust", do you mean unsubscribe? This is not a solution if you still want to see relevant updates to a discussion. Sure, my comment won't stop anything, but an uncommented chain of "+1" comments also gives folks the impression that this is the correct way to vote on an issue, when it's not.

**Reply by cantecim** [reply]
@vintagesucks Yes exactly, I guess, soon, I will have to unsubscribe to the discussion too. Sure we need another feature request here to adjust subscription in a more convenient way. Those new +1 responses are not going to stop until someone handles the discussion. Fortunately we have a lot of more upvote than the comments so I assume majority of the folks knows the difference between them. Having said that, this is not solving future +1 responses

**Reply by cantecim** [reply]
I mean, that's sad but if we think about current UX. Discussion feature is making a lot people to use Upvote and forget approach

**Comment by swebber-upngo** [comment]
I want something similar which is to either set a default strategy, or enforce a different merge strategy per branch.

I need to squash into main - this ties the GH issue # to the GH Pr # to the commit log
I need to rebase into dev, beta, production - these PRs trigger automatic releases to each environment, and I want unsquashed release history in these branches - the commits are always pulled from main where they were previously squashed.

**Comment by kasandravtv** [comment]
Been missing these features for many years.

It's very common to have a base branch that automatically deploys to a dev environment, which is then merged up regularly to another branch that automatically deploys to a prod environment. In those cases, you'd likely want only ff/rebase merges going to the prod branch, but you want different strategies for merging into the base branch. This very common use case is currently not possible to enforce.

**Comment by asierra-jetty** [comment]
+1

**Comment by tomsnep** [comment]
+1

**Comment by Starttoaster** [comment]
+1 would love something like this where you could select a default merge type based on a pattern match for the _source_ branch. And obviously I could also see it useful if there was an ability to match a pattern for the target branch as well.

My use case for this is that we loosely follow the "git flow" style for releases with release branches, but when we make changes to the release branch, we need to then checkpoint those changes back into main. Before getting ready for a release, we also contribute things to main. So we have two different "types" of pull requests that target main, that we use different merge strategies for. It would be nice if this feature supported defaulting based on the source branch, to remove a bit of potential for human error there.

**Comment by evasuarez22** [comment]
+1 

It is a functionality that is necessary in every new project that is started. In addition, on many occasions not having the ability to adjust the merge strategy depending on the base branch leads us to have to give admin permissions to most of the team so they can enable the correct strategy, which is not very advisable.

**Comment by alvarolgn** [comment]
+1
It is a very necessary feature. Having all the merge possibilities available is a risk that could be solved with defining merge strategies depending on the branch.

**Comment by Kopiczek** [comment]
+1

**Comment by asoleshimoku** [comment]
+1

**Comment by andriibrz** [comment]
+1

**Comment by jritst** [comment]
+1

**Comment by EvanSchleret** [comment]
+1

**Comment by DominicGBauer** [comment]
+1

**Comment by ecosse3** [comment]
+1

**Comment by Arielliv** [comment]
+1

**Comment by sprysyaz** [comment]
+1

**Comment by kirpichyovss** [comment]
+1

**Comment by iain17** [comment]
+1

**Comment by williamlin330088** [comment]
+1

**Comment by sebastien-savalle** [comment]
+1

**Comment by andraan** [comment]
Very helpful for large teams 👍

**Comment by OBrezhniev** [comment]
+1

**Comment by nanduri1998** [comment]
+1

**Comment by tt-cricut** [comment]
+1

**Comment by joaovinicius** [comment]
+1

**Comment by nadavsinai-philips** [comment]
+1

**Comment by jqueguiner** [comment]
+1

**Comment by rin4ik** [comment]
+1 Please make an option to change the default merge strategy.

**Comment by ntt2k** [comment]
+1

please implement this feature

**Comment by SergeyBarskiyTyler** [comment]
+1

**Comment by dertdog** [comment]
+1

**Comment by ricardogoncalves** [comment]
+1

**Comment by eyadmba** [comment]
Guys can you stop adding +1's please.
People are subscribed to this thread to get meaningful updates, not your "+1" notification. There's a lot of people here. Be considerate.

Reactions to the post are how it gets ranked [here](https://github.com/orgs/community/discussions/categories/pull-requests), so react to the post, or bug GitHub on LinkedIn or Twitter or something, idk.
Your +1 messages aren't doing anything but blasting everyone's notifications. Please.

**Comment by eyadmba** [comment]
I was wrong. The reactions to the post are **_not_** how it gets ranked. It seems to be based on latest activity.
So I guess let's go back to posting +1's so that GitHub can keep ignoring us politely again. I have no clue how they prioritize their work, but it ain't cool.

Anyway, lets go nuts again!
I'll also unsubscribe from this thread that I truly do care about ¯\\__(ツ)_/¯

**Reply by katylava** [reply]
github i am disappoint

**Reply by AlanGreene** [reply]
@eyadmba Sort by latest activity is just the default when you load the discussions. You can change the sort order.

Choose one of the 'Top' options to sort by upvotes: 
<img width="415" alt="image" src="https://github.com/user-attachments/assets/8b535417-ec16-40af-91a6-0371584d680b">

For example, sorting by 'Top: All':
<img width="1432" alt="image" src="https://github.com/user-attachments/assets/fe836b2c-1c14-4426-8347-867da5461a28">

**Comment by Waxo** [comment]
+1 it is then ¯\_(ツ)_/¯

**Comment by ecosse3** [comment]
+1. I need this.

**Comment by resthomas** [comment]
+1

**Comment by DavidIbarr** [comment]
+1 absolutely!

**Comment by Hamaster2** [comment]
akegaviar:patch-1

**Comment by gopivalleru** [comment]
+1

**Comment by shubham-k-garg** [comment]
+1

**Comment by ifoutz** [comment]
+1

**Comment by CharlesMiddleton** [comment]
+1

**Comment by BryanADavis** [comment]
+1

**Comment by Liskan** [comment]
+1

**Comment by Deltap3** [comment]
+1

**Comment by emul-docler** [comment]
+1

**Comment by bankyadam-sx** [comment]
+1

**Comment by gopivalleru** [comment]
+100

**Comment by laurogripa** [comment]
+1000

**Comment by PixelsEngineer** [comment]
Make totally sense, +1

**Comment by rajeshbolisetty** [comment]
+1

**Comment by beingshahidali** [comment]
Hi @Snailedlt,

Great point! Setting a default merge strategy for organizations or repositories would significantly improve workflow and minimize errors. Here are a few key suggestions:

Repository Settings: Allow admins to set a default merge strategy.
Branch Restrictions: Enable different defaults for various branches.
Org Settings: Apply a default for all new repositories.

Highlighting the default in the merge dropdown (perhaps with color or an emoji) would provide clear guidance to users.

These changes could streamline merges and enhance the GitHub experience!

**Comment by jmoore240** [comment]
+1

**Comment by aberenda-optifino** [comment]
+1

**Comment by mehulmina064** [comment]
+1

**Comment by radiomime** [comment]
+1

**Comment by Dev-Paulo-Henrique** [comment]
+1 #1

**Comment by krisztamas** [comment]
Two years without answer... +1

**Comment by DzmVasileusky** [comment]
+1

I think it would be pretty simple to implement since UI already has `squash and rebase` methods enforcing via branch protection.
We want linear strategy for the `develop` branch with `squash` but avoid noise and conflicts when merging hotfixes created from `master` in both branches

**Comment by neilsoult** [comment]
Please papa microsoft, before Lina Khan busts your monopoly 
+1

**Comment by Myera2411** [comment]
> # Default Merge Strategy
> ## Why we need it
> In our GitHub organization we have a workflow where the preferred merge strategy is `Squash and merge`. Whenever we merge into master, we usually use this strategy. We would like to be able to set this as the default merge strategy, without removing the options for `Create a merge commit` and `Rebase and merge` for special occasions. Currently the only way to set `Squash and merge` is to disable the other options. This is problematic, since we sometimes need to merge using other merge strategies in special cases, for example if the PR closes multiple issues, and we want one commit per issue it solves. Another thing that makes this hard is that currently GitHub selects the merge strategy based on what you selected ln your last PR. This means that when an unusual option is selected, the user needs to remember to change it back to the "default" so that he/she doesn't accidentally merge with the wrong strategy on their next merge. This is a big source of user errors, and can lead to multiple hours spent trying to fix the (sometimes irreversible) mistakes done by choosing the wrong merge strategy. Instead of having the default be the last selected strategy, it should be possible for repository admins to select which merge strategy is defaulted to.
> 
> ## Suggested Implementations
> 1. Add an option in repository settings to choose the default merge strategy.
> 2. Add an option in branch resteictions, to choose the default merge strategy.
> 3. Add an option in the GitHub organization settings to choose the default merge strategy for any newly created repositories.
> 
> Additionally, if the GitHub UX team finds it useful, it would be nice to have the default merge strategy highlighted in the merge button dropdown. Either with a different background color, or an associated emoji to show if the merge strategy is a default in the org, repo or branch.
> 
> ## Implementations and other relevant links
> ### Relevant Links
> * [Merge methods per branch #8196](https://github.com/orgs/community/discussions/8196)
> * [Organization defaults/templates for repositories (GitHub Enterprise) #13991](https://github.com/orgs/community/discussions/13991)
> * [GitHub Community“Squash and Merge” as default Pull Request merge option](https://github.community/t/squash-and-merge-as-default-pull-request-merge-option/1031)
> * [StackOverflow | Is there a way to specify a default option for merging a pull request in GitHub?](https://stackoverflow.com/q/49203040/12206312)
> 
> ### Implementations by GitHub alternatives
> [Bitbucket Cloud](https://bitbucket.org/blog/fast-forward-merges-bitbucket-cloud-default-like) (My favourite) <img alt="Screen-Shot-2018-06-06-at-3 53 53-PM" width="826" src="https://user-images.githubusercontent.com/43886029/179179198-63e375b8-462b-4a99-ba42-11f8cfe17322.png">
> 
> [Bitbucket Server](https://confluence.atlassian.com/bitbucketserver/pull-request-merge-strategies-844499235.html#Pullrequestmergestrategies-Configuremergestrategiesforanindividualrepository) <img alt="Screen-Shot-2018-06-06-at-3 57 11-PM" width="698" src="https://user-images.githubusercontent.com/43886029/179179244-a48102b9-a6c9-430b-85ef-18a46f758238.png">
> 
> [Gitlab](https://gitlab.com/gitlab-org/gitlab/-/issues/17613) ![image](https://user-images.githubusercontent.com/43886029/179180180-ee4869c2-0b8c-4709-beda-b2689047df0a.png)
> 
> [Gitea](https://github.com/go-gitea/gitea/issues/17741) ![142775768-97a8368f-531e-4b71-9ef6-39c56147f9f9](https://user-images.githubusercontent.com/43886029/179187816-20a55536-26be-44af-84f2-6d2e51df00b8.png)

**Comment by GSudeendra** [comment]
+1

**Comment by nam-truong-le** [comment]
+1

**Comment by DominicGBauer** [comment]
+1

**Comment by ssyberg** [comment]
+1

**Comment by patrick-knight** [comment]
TL;DR - Were working on it. 

We are going to address this with [repository rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets) as the way to limit which merge methods are available per branch.

More details here in the other discussion post https://github.com/orgs/community/discussions/10809#discussioncomment-10329765

**Reply by Snailedlt** [reply]
Awesome!
Would be nice to see another option for default branch too though. Some branches you want to squash and merge most of the time, but on occasion you might want to just merge normally.

For example you usually wanna have 1 PR per feature. In which case you name the PR with the name of the feature and squash merge it. But sometimes you may have features that depend on each other, where you want to merge them at the same time, but still keep the commits separate so that the automated changelog and release notes show all of the new features.

If this is the case for your repo you wouldn't want to deny normal merging, but you would like to have them make a conscious decision to do anything other than squash merging.

This is where the default merge type would be really handy!

**Reply by patrick-knight** [reply]
This new PR merge method rule is rolling out now in preview. 


Please direct your feedback about the behavior and configuration in this new discussion post: https://github.com/orgs/community/discussions/146284

**Reply by sedsedus** [reply]
@patrick-knight 
Repository rules are great, however, this doesn't address the main issue highlighted by the OP - which is the error-prone nature of the default selection (remembered from the last actual merge). 
In my project, we sometimes have to merge (to the same branch) with squash, and sometimes without. So having a 1/0 kind of switch will not help us at all.

**Reply by merusso** [reply]
@Snailedlt , I suggest that you unmark this answer for now. If it's "answered", people will think this feature is now available, but the behavior you're asking for isn't yet available.

The GitHub feature enables a repo to restrict PR merge types. It doesn't offer a means to set a default merge type.

Thoughts?

**Reply by Snailedlt** [reply]
@merusso agreed. unmarked

**Reply by Snailedlt** [reply]
@patrick-knight I see you've moved to a different company, but do you happen to know who currently leads the relevant team for this feature?

It's one of the most requested features on GitHub, and while the solution your team made is good, it doesn't really cover the needs described in the issue.


It would be awesome if we could get this feature added ASAP!

**Comment by MichaelChirico** [comment]
I'm not 100% sure this is the exact issue for this feedback (since this seems to be mainly about "rebase" vs. "squash"), but it's at least similar.

One pain point I have is that my merge strategy (`ort`) on local works perfectly to resolve conflicts in our NEWS file, but GitHub's merge strategy always finds conflicts.

This is a pain because resolving conflicts wipes out reviewer Approvals, whereas conflict-free merges do not.

That means that on local, I can easily `merge` with no conflict, but on GitHub I have to invalidate review to proceed (generating lots of reviewer toil). And of course pushing the conflict-free merge from local _also_ invalidates the review.

This is especially painful if I go on a productive streak and push 5-6 PRs, each of which touch the NEWS. It generates a tedious dance where reviewers have to continuously re-Approve PRs (n choose 2 --> a quadratically increasing number of approvals).

**Comment by MaxBeauchemin** [comment]
+1

**Comment by bleadof** [comment]
+1

**Comment by allan-wright-pushpress** [comment]
+1

**Comment by tdubourgR** [comment]
Not sure what reasons there would be _not_ to implement this. This is currently causing a significant amount of wasted time in my team, not to have a selectable default, due to various devs having different workflows.

**Comment by conditionsofexistence** [comment]
+1

**Comment by jonas-singe-bruker** [comment]
+1

**Comment by matthieu-vergne** [comment]
+1

**Comment by wasifali758595** [comment]
In GitHub (and similar platforms like GitLab/Bitbucket), the merge strategy determines how pull requests (PRs) are merged into a target branch. By default, GitHub allows multiple strategies, and admins can configure which ones are enabled or set as the default.

Available Merge Strategies

Merge Commit

Creates a merge commit that brings the PR into the base branch.

Preserves full commit history.

Default in many repos.

Squash and Merge

Squashes all commits from the PR into a single commit on the base branch.

Keeps history clean and linear.

Useful if PRs have noisy commits.

Rebase and Merge

Replays commits from the PR onto the tip of the base branch without creating a merge commit.

Results in a linear history but preserves individual commits.

Setting the Default Merge Strategy

Per Repository:

Go to Repository Settings → General → Pull Requests → Merge Button.

Enable/disable strategies (Merge Commit, Squash, Rebase).

Select a default merge method (the one pre-selected in PRs).

Per Organization:

Org admins can configure repository defaults under Organization Settings → Repository Defaults.

These defaults apply to new repositories created in the org.

Per Branch:

GitHub doesn’t let you set a merge strategy per branch directly.

But you can enforce this behavior with:

Branch Protection Rules + requiring linear history (forces squash or rebase).

Or by disabling unwanted merge methods at the repo level.

Best Practices

Squash & Merge: Good for teams that want one commit per feature/PR → cleaner history.

Merge Commit: Best if you want to preserve context of feature branches.

Rebase & Merge: Best if you want linear history and to preserve individual commits.

✅ Summary:
The default merge strategy is configurable per repository (via settings) and can be standardized per organization using defaults. For per branch, you enforce via branch protection rules (linear history), not by direct default selection.

**Reply by merusso** [reply]
There is no way to configure a default merge strategy yet. You can turn off certain merge strategies entirely, but you can't set a repo-wide default strategy.

**Reply by samus-aran** [reply]
Hi @wasifali758595 , 

We’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. Thank you for helping us maintain an authentic and beneficial space for everyone.

**Comment by matthieu-vergne** [comment]
> Per Repository:
> 
> Go to Repository Settings → General → Pull Requests → Merge Button.
> 
> Enable/disable strategies (Merge Commit, Squash, Rebase).
> 
> Select a default merge method (the one pre-selected in PRs).

1) I see no "Merge Button" in the PR settings of my repo.
2) Enabling/disabling was explicitly stated to not solve the issue.
3) The preselection in PRs is not a default, it is the next one used until it get changed

**Reply by mrmoxon** [reply]
Yeah this still hasn't been resolved. Only option is to disallow traditional merge commits.

**Comment by bengaudiosi-toast** [comment]
+1

**Comment by Wyntuition** [comment]
+1

**Comment by nex-jesusjuarez** [comment]
+1

**Comment by nonemec-havok** [comment]
Definitely, please! I typically do squash merges but need different strategies in some cases. Afterwards I regularly forget that on the next PR and accidentally do a non-squash without good reason. The "remember last merge per user" behavior is really bad UI design!

**Comment by andrewaarestad** [comment]
We would appreciate a "default merge strategy" feature - sounds like it's not well-aligned with GitHub's product roadmap though. Maybe limiting merge strategies based on user/role? 

Basically we want to have our devs use squash commits when merging PRs to the default branch, but we want to do backmerge PRs that use merge strategy.

**Reply by evanrasmussen9** [reply]
+1 to this comment. We would love to have the ability for users in group X to be able to squash or merge, but users in group y only be able to merge.

**Comment by rajeev-vishnu** [comment]
+1

**Comment by carlosalban** [comment]
+1

**Comment by Golden00885** [comment]
+1 for this feature, can't believe it does not exist yet!! :)

**Comment by leouo** [comment]
+1

**Comment by mayankshah1607** [comment]
+1

**Comment by Myera2411** [comment]
Dear Jørgen, Thank you for reaching out! I appreciate your insights regarding
the feature request. While I may not have the latest updates on team
leadership, I'm glad to hear that there's strong interest in enhancing our
offerings on GitHub.  Your enthusiasm for this feature is contagious, and I
genuinely hope that the necessary updates and improvements can be prioritized
soon. Please feel free to share any other thoughts or needs you might have.
Best regards,  Nurrul Era Hamzie Binti Hamzah  
  

-  

  

On 5 May 2026 3:57:37 am Jørgen Kalsnes Hagen ***@***.***&gt;
wrote:  

> ***@***.***(https://github.com/patrick-knight) I see you've moved to a
different company, but do you happen to know who currently leads the relevant
team for this feature?

>

> It's one of the most requested features on GitHub, and while the solution
your team made is good, it doesn't really cover the needs described in the
issue.

>

> It would be awesome if we could get this feature added ASAP!

>

> --  
> Reply to this email directly, [view it on
GitHub](https://github.com/orgs/community/discussions/20583#discussioncomment-16809332),
or [unsubscribe](https://github.com/notifications/unsubscribe-
auth/BMC2G5OMNKPMMBF6QTF23XD4ZDY3DAVCNFSM53U3MPFKU5DIOJSWCZC7NNSXTOSENFZWG5LTONUW63SDN5WW2ZLOOQ5TCNRYGA4TGMZS).  
> Triage notifications on the go with GitHub Mobile for
[iOS](https://apps.apple.com/app/apple-store/id1477376905?ct=notification-
email&mt=8&pt=524675) or
[Android](https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign
%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub).  
> You are receiving this because you
commented.![](https://github.com/notifications/beacon/BMC2G5IQXCGYBYBN54DGQFD4ZDY3DBFCNFSM53U3MPFKYY3PNVWWK3TUL52HS4DFWFCGS43DOVZXG2LPNZBW63LNMVXHJKTDN5WW2ZLOORPWSZGOAEAH25FGOJSWC43PN2TWG33NNVSW45A.gif)Message
ID: &lt;community/community/repo-
***@***.***&gt;

**Comment by splouisliu** [comment]
+1 Bruh it's been 4 years how does this still not exist yet...

**Comment by jo-takao** [comment]
>Another thing that makes this issue more serious is that GitHub automatically selects the merge
strategy based on what the user chose in their previous PR.

This is not just a matter of convenience or UX preference — it is a risky behavior that can easily
lead to accidental merges with the wrong method, especially when switching between repositories
or branches with different merge policies.

Because of this, I believe this behavior should be treated not as a “nice-to-have improvement,”
but as something that needs to be corrected to prevent workflow-breaking mistakes.

Thanks for paying attention to [my comment](https://github.com/orgs/community/discussions/146284#discussioncomment-17233344). I appreciate it.

There are tons of threads like this lately. I really hope we can see meaningful progress toward a solution soon.

**Comment by colpatch-waving** [comment]
+1 Please implement this :pray:
